core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java

```diff
@@ -124,16 +125,54 @@ public String generateLoadBalancerEndpoint(Stack stack) {
         }
     }
 
+    /*
+     *  follows these peferences:
+     *   public fqdn
+     *   public DNS
+     *   public IP
+     *   private FQDN
+     *   private DNS
+     *   private IP
+     */
     public String getLoadBalancerUserFacingFQDN(Long stackId) {
         Set<LoadBalancer> loadBalancers = loadBalancerPersistenceService.findByStackId(stackId).stream()
-            .filter(lb -> StringUtils.isNotEmpty(lb.getFqdn()))
-            .collect(Collectors.toSet());
-        LoadBalancer loadBalancer = loadBalancers.stream()
-            .filter(lb -> LoadBalancerType.PUBLIC.equals(lb.getType()))
-            .findAny()
-            .orElseGet(() -> loadBalancers.stream().findAny().orElse(null));
+                .filter(lb -> StringUtils.isNotEmpty(lb.getDns()) ||
+                        StringUtils.isNotEmpty(lb.getIp()) || StringUtils.isNotEmpty(lb.getFqdn()))
+                .collect(Collectors.toSet());
+        if (loadBalancers.isEmpty()) {
+            return null;
+        }
+        Iterator<LoadBalancer> iterator = loadBalancers.iterator();
+        LoadBalancer best = iterator.next();
+        while (iterator.hasNext()) {
```

Comment:
I'm not entirely sure what changing the original stream() to this iterator adds to the logic. It looks like what it's doing is iterating over the set of load balancers to get the PUBLIC load balancer. But that's already what the stream() was doing:
```
LoadBalancer loadBalancer = loadBalancers.stream()
    .filter(lb -> LoadBalancerType.PUBLIC.equals(lb.getType()))
    .findAny()
    .orElseGet(() -> loadBalancers.stream().findAny().orElse(null));
```
That'll look at the set, select the PUBLIC load balancer if one exists in the set, and select the first load balancer in the set if there isn't a PUBLIC one. So I think you can just take the load balancer is selects here and pass that to `getBestAddressable`.